File: Kalman_Filter_UKF.m
% 没有设置随机种子，所以每次运行代码时都会生成不同的随机噪声。

% 清空变量和命令窗口
clear;
clc;

% 生成模拟数据
dt = 1; % 时间步长
steps = 100; % 时间步数
x = zeros(1, steps); % 真实状态
z = zeros(1, steps); % 测量数据
x(1) = 0; % 初始状态
Q = 0.01^2; % 系统噪声协方差
R = 0.1^2; % 测量噪声协方差

% 系统动态方程和测量方程
f = @(x) x + sin(x)*dt; % 非线性系统方程
h = @(x) x; % 测量方程

% 生成真实状态和测量数据
for t = 2:steps
    x(t) = f(x(t-1)) + sqrt(Q)*randn();
    z(t) = h(x(t)) + sqrt(R)*randn();
end

% 初始化 UKF 参数
nx = 1; % 状态维度
nz = 1; % 测量维度
alpha = 1e-3;
beta = 2;
kappa = 0;
lambda = alpha^2*(nx+kappa)-nx;
wm = [lambda/(nx+lambda) repmat(1/(2*(nx+lambda)), 1, 2*nx)]; % 权重
wc = wm;
wc(1) = wc(1) + (1 - alpha^2 + beta);
Xs = zeros(nx, 2*nx+1); % sigma 点
P = Q; % 初始协方差矩阵
xhat = zeros(1, steps); % 估计状态
xhat(1) = 0; % 初始状态估计

% UKF 滤波过程
for t = 2:steps
    % 计算 sigma 点
    X = sqrtm((nx+lambda)*P);
    Xs = [xhat(t-1) xhat(t-1)+X xhat(t-1)-X];
    
    % 预测
    Xs_pred = arrayfun(f, Xs);
    xhat_pred = wm*Xs_pred';
    P_pred = (wc .* (Xs_pred - xhat_pred)) * (Xs_pred - xhat_pred)';
    P_pred = P_pred + Q;
    
    % 更新
    Zs = arrayfun(h, Xs_pred);
    zhat = wm*Zs';
    Pzz = (wc .* (Zs - zhat)) * (Zs - zhat)';
    Pzz = Pzz + R;
    Pxz = (wc .* (Xs_pred - xhat_pred)) * (Zs - zhat)';
    K = Pxz / Pzz;
    xhat(t) = xhat_pred + K*(z(t) - zhat);
    P = P_pred - K*Pzz*K';
end

% 绘图
figure;
plot(1:steps, x, 'LineWidth', 1.5); hold on;
plot(1:steps, z, 'o', 'MarkerSize', 2);
plot(1:steps, xhat, 'LineWidth', 1.5);
legend('真实状态', '测量', 'UKF 估计');
grid on;
xlabel('时间步');
ylabel('状态');
title('无迹卡尔曼滤波器 (UKF) 状态估计');

File: Kalman_Filter_UKF.py
import numpy as np
import matplotlib.pyplot as plt
from filterpy.kalman import UnscentedKalmanFilter, MerweScaledSigmaPoints
plt.rcParams['font.sans-serif'] = ['SimHei']
plt.rcParams['axes.unicode_minus'] = False

# 定义非线性函数
def f(x, dt):
    return x + np.sin(x) * dt

# 定义 UKF 的初始状态和噪声
x0 = np.array([0.])
Q = np.diag([0.01**2])   # 系统噪声的协方差
R = 0.1**2               # 测量噪声的协方差

# 定义 Unscented Kalman Filter
points = MerweScaledSigmaPoints(1, alpha=1e-3, beta=2, kappa=0)
ukf = UnscentedKalmanFilter(dim_x=1, dim_z=1, dt=1,
                            fx=f, hx=lambda x: x,
                            points=points)
ukf.x = x0
ukf.Q = Q
ukf.R = R

# 生成模拟数据
dt = 1
steps = 100
x = np.zeros((steps, 1))
x[0] = x0
for t in range(1, steps):
    x[t] = f(x[t-1], dt) + np.sqrt(Q) * np.random.randn(1, 1)

# 添加测量噪声
z = x + np.sqrt(R) * np.random.randn(steps, 1)

# 使用 UKF 进行滤波
ukf_xs = np.zeros((steps, 1))
for t in range(steps):
    ukf.predict()
    ukf.update(z[t])
    ukf_xs[t] = ukf.x

# 绘制结果
plt.figure()
plt.plot(x, label='真实值')
plt.plot(z, 'o', label='测量值')
plt.plot(ukf_xs, label='UKF估计值')
plt.legend()
plt.show()
